ntp.go

```diff
@@ -407,78 +409,57 @@ func parseTime(m *msg, recvTime ntpTime) *Response {
+	}
+	return rtt
 }
 
-func offset(t1, t2, t3, t4 ntpTime) time.Duration {
-	// local offset equation (https://tools.ietf.org/html/rfc5905#section-8)
-	//   T1 = client send time
-	//   T2 = server receive time
-	//   T3 = server reply time
-	//   T4 = client receive time
-	//
-	// Local clock offset t:
-	//   t = ((T2-T1) + (T3-T4)) / 2
-	a := t2.Time().Sub(t1.Time())
-	b := t3.Time().Sub(t4.Time())
+func offset(org, rec, xmt, dst ntpTime) time.Duration {
+	// local clock offset
+	//   offset = ((rec-org) + (xmt-dst)) / 2
+	a := rec.Time().Sub(org.Time())
+	b := xmt.Time().Sub(dst.Time())
 	return (a + b) / time.Duration(2)
 }
 
-func rootDistance(rtt, delay, disp time.Duration) time.Duration {
-	// RFC5905 suggests more strict check against _peer_ in fit(), that
-	// root_dist should be less than MAXDIST + PHI * LOG2D(s.poll).
-	// MAXPOLL is 17, so it is approximately at most (1s + 15e-6 * 2**17) =
-	// 2.96608 s, but MAXDIST and MAXPOLL are confugurable values in the
-	// reference implementation, so only MAXDISP check has hardcoded value
-	// in Validate().
+func rootDistance(rtt, rootDelay, rootDisp time.Duration) time.Duration {
+	// The root distance is:
+	// 	the maximum error due to all causes of the local clock
+	//	relative to the primary server. It is defined as half the
+	//	total delay plus total dispersion plus peer jitter.
+	//	(https://tools.ietf.org/html/rfc5905#appendix-A.5.5.2)
 	//
-	// root_dist should also have following summands
-	// + Dispersion towards the peer
-	// + jitter of the link to the peer
-	// + PHI * (current_uptime - peer->uptime_of_last_update)
-	// but all these values are 0 if only single NTP packet was sent.
-	if rtt < 0 {
-		rtt = 0
-	}
-	return (rtt+delay)/2 + disp
-}
-
-func causalityViolation(rtt, offset time.Duration) time.Duration {
-	// NTP query has four timestamps for consecutive events: T1, T2, T3
-	// and T4. T1 and T4 use the local clock, T2 and T3 the server clock.
-	// RTT    = (T4 - T1) - (T3 - T2)     =   T4 - T3 + T2 - T1
-	// Offset = (T2 + T3)/2 - (T4 + T1)/2 = (-T4 + T3 + T2 - T1) / 2
-	// => T2 - T1 = RTT/2 + Offset && T4 - T3 = RTT/2 - Offset
-	// If system wall-clock is synced to NTP-clock then T2 >= T1 && T4 >= T3.
-	// This check may be useful against chrony NTP daemon as it starts
-	// relaying sane NTP clock before system wall-clock is actually adjusted.
-	violation := rtt / 2
-	if offset > 0 {
-		violation -= offset
-	} else {
-		violation += offset
-	}
-
-	if violation < 0 {
-		return -violation
+	// In the reference implementation, it is calculated as follows:
+	//	rootDist = max(MINDISP, rootDelay + rtt)/2 + rootDisp
+	//			+ peerDisp + PHI * (uptime - peerUptime)
+	//			+ peerJitter
+	// For an SNTP client which sends only a single packet, most of these
+	// terms are irrelevant and become 0.
+	totalDelay := rtt + rootDelay
+	if totalDelay < minDispersion {
+		totalDelay = minDispersion
 	}
-	return time.Duration(0)
+	return totalDelay/2 + rootDisp
 }
```

Comment:
It does not handle negative RTT that may theoretically happen if clock frequency is really mad.